base/stream.jl

```diff
@@ -942,19 +942,20 @@ function readbytes!(s::LibuvStream, a::Vector{UInt8}, nb::Int)
     if bytesavailable(sbuf) >= nb
         nread = readbytes!(sbuf, a, nb)
     else
-        initsize = length(a)
-        newbuf = PipeBuffer(a, maxsize=nb)
-        newbuf.size = newbuf.offset # reset the write pointer to the beginning
+        # TODO: Here, we don't read directly into `a`, but read to
+        # a new buffer, and then copy to `a`
+        newbuf = PipeBuffer(; maxsize=nb)
```

Comment:
This seems to be a regression?

Reply:
This here is the same as the next three. `IOBuffer` owns the underlying data once it has been written to. However, the previous code here attempted to save an allocation by wrapping an existing vector in an `IOBuffer`, and then writing to the vector by writing to the `IOBuffer`, before using the vector.
This is invalid, because the `IOBuffer` owns the vector.
To see why it's an issue, it means that this code relies on undocumented and possibly changing details about how the `IOBuffer` reallocates the buffer. For example, if we ever decide to reallocate the memory of `IOBuffers` pre-emptively if enough data is written to them, this, seemingly unrelated code will break.

Reply:
This was the original purpose of IOBuffer though, not the other way around. The original intent was supporting zero-copy IO specifically for this file, while it just happens to also be a fairly usable IO abstraction also.

> This is invalid, because the IOBuffer owns the vector.

That is true, then we destroy the IOBuffer at the end (_unsafe_take!) and rebuild the vector (_take!) to transfer ownership back to the Vector so that at any given time there is only one owner of the buffer